### ISSUE TITLE
Custom preprocessing wrapper and Target

### ISSUE THREAD

**Issue by xenmind** [mlr-org/mlr#1770]
Hi,

How might I access a test set's target values (or the test set's index values) in a custom preprocessing wrapper predict function please?

Thanks,
Gavin.

**Comment by larskotthoff** [comment]
The target values aren't passed to the `predictLearner` function, which is what you would have to overwrite here. In principle, you could keep the data somewhere in the global environment and access the target values through that.

What are you trying to achieve?

**Comment by xenmind** [comment]
I'm thinking of applying a single parameter function to the target before learning. So the initial target is modified to create a filtered target. I was hoping to optimize this function by creating a preprocessing wrapper. Re: the global env. method, how do I get the index values of the test or train set, so that I'm pulling the right data from the total set?

**Comment by larskotthoff** [comment]
It sounds like you would want to do that on the raw data before passing it to mlr. What kind of filtering do you have in mind?

**Comment by xenmind** [comment]
True, but then how do I optimize the parameter value so that the final optimized target is the one that minimises RMSE? The filter is a type of low pass filter.

**Comment by larskotthoff** [comment]
Hmm, that sounds weird. So you want to basically throw away data points for which you don't get good predictions? What's the application?

**Comment by xenmind** [comment]
The filtering doesn't remove data, it adjusts them to smooth the signal, think of a moving average. I'm trying to predict t+1 of a smoothed signal: Y -> Ysmooth -> Ysmooth(t+1). I have found an issue in that when the test set is sampled, the function doesn't generate the same values as when I use it on unsampled data before inputting to mlr, but I was experimenting to see if this concept might still work.

**Comment by larskotthoff** [comment]
Ah, I see. So you want to do time-series forecasting? And the parameter of the low-pass filter should probably be set to a single set of values for both training and testing, so one filter applied to the entire data.

It sounds like a custom optimization method (which internally calls one of the existing ones) would be a good way to do this. The problem with a custom preprocessing method or learner is that you never get full access to the data.

The easiest technically is certainly to simply optimize the filter parameters outside of mlr, i.e. tune for different data sets. How many different parameter values for the filter are you considering?

**Comment by xenmind** [comment]
Haha sorry I should have been clearer on what I'm doing. A custom optimization method sounds beyond my abilities, doesn't look like there is a convenient wrapper interface for this. I don't have many parameter values, so running my mlr code in a for loop over a range is surely the simplest thing to do. Thanks so much for your help!

**Comment by larskotthoff** [comment]
It sounds like you definitely want a separate unsmoothed test set though to avoid having a very good model of a function that's smoothed so much that it doesn't help you with the original.

**Comment by xenmind** [comment]
Yes, my initial thought on doing this was that it's a great way to overfit the data, so I have a separate holdout set to verify. Cheers.
